syntax = "proto3";

package aico.modelservice;

option go_package = "github.com/boeni-industries/aico/proto/modelservice";
option java_package = "industries.boeni.aico.proto.modelservice";
option java_multiple_files = true;

import "google/protobuf/timestamp.proto";

// ============================================================================
// REQUEST MESSAGES
// ============================================================================

// Health check request
message HealthRequest {
  // Empty - no parameters needed
}

// Conversation completions request
message CompletionsRequest {
  string model = 1;
  repeated ConversationMessage messages = 2;
  bool stream = 3;
  optional double temperature = 4;
  optional int32 max_tokens = 5;
  optional double top_p = 6;
  optional string system = 7;
}

// Available models request
message ModelsRequest {
  // Empty - no parameters needed
}

// Model info request
message ModelInfoRequest {
  string model = 1;
}

// Embeddings request
message EmbeddingsRequest {
  string model = 1;
  string prompt = 2;
}

// NER request
message NerRequest {
  string text = 1;
}

// Sentiment analysis request
message SentimentRequest {
  string text = 1;
}

// Service status request
message StatusRequest {
  // Empty - no parameters needed
}

// Ollama management requests
message OllamaStatusRequest {
  // Empty - no parameters needed
}

message OllamaModelsRequest {
  // Empty - no parameters needed
}

message OllamaPullRequest {
  string model = 1;
}

message OllamaRemoveRequest {
  string model = 1;
}

message OllamaServeRequest {
  // Empty - no parameters needed
}

message OllamaShutdownRequest {
  // Empty - no parameters needed
}

// ============================================================================
// RESPONSE MESSAGES
// ============================================================================

// Health check response
message HealthResponse {
  bool success = 1;
  string status = 2;
  optional string error = 3;
}

// Conversation completions response
message CompletionsResponse {
  bool success = 1;
  optional CompletionResult result = 2;
  optional string error = 3;
}

// Available models response
message ModelsResponse {
  bool success = 1;
  repeated ModelInfo models = 2;
  optional string error = 3;
}

// Model info response
message ModelInfoResponse {
  bool success = 1;
  optional ModelDetails details = 2;
  optional string error = 3;
}

// Embeddings response
message EmbeddingsResponse {
  bool success = 1;
  repeated double embedding = 2;
  optional string error = 3;
}

// NER response
message NerResponse {
  bool success = 1;
  map<string, EntityList> entities = 2;
  optional string error = 3;
}

message EntityList {
  repeated string entities = 1;
}

// Sentiment analysis response
message SentimentResponse {
  bool success = 1;
  string sentiment = 2;  // "positive", "negative", "neutral"
  double confidence = 3; // confidence score 0.0-1.0
  optional string error = 4;
}

// Service status response
message StatusResponse {
  bool success = 1;
  optional ServiceStatus status = 2;
  optional string error = 3;
}

// Ollama management responses
message OllamaStatusResponse {
  bool success = 1;
  optional OllamaStatus status = 2;
  optional string error = 3;
}

message OllamaModelsResponse {
  bool success = 1;
  repeated ModelInfo models = 2;
  optional string error = 3;
}

message OllamaPullResponse {
  bool success = 1;
  optional string message = 2;
  optional string error = 3;
}

message OllamaRemoveResponse {
  bool success = 1;
  optional string message = 2;
  optional string error = 3;
}

message OllamaServeResponse {
  bool success = 1;
  optional string message = 2;
  optional string error = 3;
}

message OllamaShutdownResponse {
  bool success = 1;
  optional string message = 2;
  optional string error = 3;
}

// ============================================================================
// SHARED DATA STRUCTURES
// ============================================================================

// Conversation message for completions
message ConversationMessage {
  string role = 1;     // "system", "user", "assistant"
  string content = 2;
}

// Completion result
message CompletionResult {
  string model = 1;
  google.protobuf.Timestamp created_at = 2;
  ConversationMessage message = 3;
  bool done = 4;
  optional int64 total_duration = 5;
  optional int64 load_duration = 6;
  optional int64 prompt_eval_count = 7;
  optional int64 prompt_eval_duration = 8;
  optional int64 eval_count = 9;
  optional int64 eval_duration = 10;
}

// Model information
message ModelInfo {
  string name = 1;
  string model = 2;
  google.protobuf.Timestamp modified_at = 3;
  int64 size = 4;
  string digest = 5;
  optional ModelDetails details = 6;
}

// Detailed model information
message ModelDetails {
  string parent_model = 1;
  string format = 2;
  string family = 3;
  repeated string families = 4;
  int64 parameter_size = 5;
  int64 quantization_level = 6;
}

// Service status information
message ServiceStatus {
  string version = 1;
  bool ollama_running = 2;
  string ollama_version = 3;
  int32 loaded_models_count = 4;
  repeated string loaded_models = 5;
}

// Ollama status information
message OllamaStatus {
  bool running = 1;
  string version = 2;
  string host = 3;
  int32 port = 4;
  repeated string loaded_models = 5;
}
